### DISCUSSION TITLE
Got Nan grad in backpropagation

### DISCUSSION THREAD

**Discussion by Parsifal133** [kornia/kornia#2389] [Questions] [answered]
When I'm using SSIM-LOSS to calculate the difference of feature maps, I get normal loss, but when the loss is backpropagated, I get NAN gradient. It should be noted that I am using yolov5 v3.1, and this problem does not exist when I switch to yolov5 v6.0. This means that the code of ssim should be no problem. 
I then compared the feature maps of the v6.0 and v3.1 versions, and the difference  is that the feature map of v3.1 has more zero values. What are the possible reasons for this problem? I urgently need to use ssim loss on yolov5 v3.1

**Comment by edgarriba** [comment]
You should mention that to yolov5 repo — we are unrelated orgs

**Comment by edgarriba** [comment]
FYI, we are adding soon an ObjectDetection API — https://github.com/kornia/kornia/pull/2363
Which we are still validating with the different stakeholders of the project.
Besides, what is your exact use case to train with ssim loss ?

**Reply by Parsifal133** [reply, marked as answer]
> FYI, we are adding soon an ObjectDetection API — #2363 Which we are still validating with the different stakeholders of the project. Besides, what is your exact use case to train with ssim loss ?

Problem solved. I use ssim for yolov5 Neck feature distillation.
Without any processing, the feature maps from yolov5-v3.1 neck contain a large number of zero values. These feature maps can calculate ssim loss normally, but cause NAN gradients when backpropagating.
I then propagate these feature maps in sigmoid before loss calculating , and get naomal gradients when backpropagating.
In my experience, a large number of zero values can be the cause of NAN gradients in backpropagation. For reference only.
